File: src/dompa.ts
import { Effect, Ref, Context, Option } from "effect";

/**
 * State for holding the HTML string.
 */
class HtmlState extends Context.Tag("HtmlState")<
  HtmlState,
  Ref.Ref<string>
>() {}

/**
 * -----------------------------------------------------------------
 * IR Node
 * -----------------------------------------------------------------
 */
type IrNodeProps = {
  name: string;
  coords: [number, number];
  children: IrNode[];
};

/**
 * An IR Node (intermediate representation node) is a node that collects
 * crude information such as the node name, and its coords. Based on that
 * information we also give it children, if it has any, based on coords
 */
class IrNode {
  name: string;
  coords: [number, number];
  children: IrNode[];

  constructor({ name, coords, children }: IrNodeProps) {
    this.name = name;
    this.coords = coords;
    this.children = children;
  }
}

class IrNodesState extends Context.Tag("IrNodesState")<
  IrNodesState,
  Ref.Ref<IrNode[]>
>() {}

/**
 * -----------------------------------------------------------------
 * Node
 * -----------------------------------------------------------------
 */
type Attrs = Record<string, string | true>;

type NodeProps = {
  name: string;
  attributes: Attrs;
  children: Node[];
};

/**
 * A Node is a generic HTML element that has a name, attributes
 * and children.
 */
class Node {
  name: string;
  attributes: Attrs;
  children: Node[];

  constructor({ name, attributes, children }: NodeProps) {
    this.name = name;
    this.attributes = attributes;
    this.children = children;
  }
}

/**
 * A TextNode is a Node that only contains plain text.
 */
class TextNode extends Node {
  value: string;

  constructor(value: string) {
    super({ name: "_text_node", attributes: {}, children: [] });
    this.value = value;
  }
}

/**
 * A VoidNode is a Node that is self-closing, and has no children.
 */
class VoidNode extends Node {}

/**
 * A FragmentNode is a Node which children replace itself.
 * This is useful if you want to replace the current node in the
 * traversal iteration with multiple nodes.
 */
class FragmentNode extends Node {}

/**
 * State for holding the nodes.
 */
class NodesState extends Context.Tag("NodesState")<
  NodesState,
  Ref.Ref<Node[]>
>() {}

/**
 * -----------------------------------------------------------------
 * Core logic
 * -----------------------------------------------------------------
 */
const voidNames = [
  "!doctype",
  "area",
  "base",
  "br",
  "col",
  "embed",
  "hr",
  "img",
  "input",
  "link",
  "meta",
  "source",
  "track",
  "wbr",
];

/**
 * An Effect that sets the end coord of the last IR Node with given `name`
 * to `coord`, but only on a last IR node which has its end coord 0, as in
 * unset.
 */
const setEndCoordOfLastIrNodeEffect = (name: string, coord: number) =>
  Effect.gen(function* () {
    const irNodesState = yield* IrNodesState;
    const irNodesValue = yield* Ref.get(irNodesState);

    //yield* Effect.fail(new Error(":)"));

    if (irNodesValue.length === 0) {
      yield* Effect.fail(new Error("No IrNode's found."));
    }

    const index = irNodesValue.findLastIndex(
      (n) => n.name === name && n.coords[1] === 0,
    );

    if (index === -1) {
      yield* Effect.fail(new Error("Could not find matching IrNode."));
    }

    yield* Ref.update(irNodesState, (irNodes) => {
      irNodes[index].coords = [irNodes[index].coords[0], coord];

      return irNodes;
    });

    yield* Effect.succeed(true);
  });

/**
 * An Effect that closes the last IR node matching `tag` with given
 * `coord` by setting it as an end coord.
 */
const closeIrNodeEffect = (tag: string, coord: number) =>
  Effect.gen(function* () {
    const name = tag
      .substring(2, tag.length - 1)
      .split(" ")[0]
      .trim();

    yield* setEndCoordOfLastIrNodeEffect(name, coord);
  });

/**
 * An Effect which creates IR nodes.
 */
const createIrNodesEffect = Effect.gen(function* () {
  const html = yield* Ref.get(yield* HtmlState);
  const irNodesState = yield* IrNodesState;
  let tagStart: number | null = null;
  let tagEnd: number | null = null;
  let textStart: number | null = null;
  let textEnd: number | null = null;

  for (let i = 0; i < html.length; i++) {
    const char = html.at(i);

    // Start of a tag, or end of a text node
    if (char === "<") {
      if (textStart !== null) {
        textEnd = i;
      }

      tagStart = i;
    }

    // If we're at the last char, and it's not the end of a tag,
    // means we're in a text node
    if (i === html.length - 1 && char !== ">") {
      textEnd = i + 1;
    }

    // End of a tag
    if (char === ">") {
      tagEnd = i + 1;
    }

    // When we have both tagStart and tagEnd, collect
    if (tagStart !== null && tagEnd !== null) {
      const tag = html.substring(tagStart, tagEnd);

      // If this is a closing tag, close the last node
      if (tag.startsWith("</")) {
        yield* closeIrNodeEffect(tag, tagEnd);
        tagStart = null;
        tagEnd = null;
        continue;
      }

      const name = tag
        .substring(1, tag.length - 1)
        .split(" ")[0]
        .trim();

      // If this is a void node, we're already at the end.
      if (voidNames.includes(name.toLowerCase())) {
        const newIrNode = new IrNode({
          name,
          coords: [tagStart, tagEnd],
          children: [],
        });

        yield* Ref.update(irNodesState, (irNodes) => {
          return [...irNodes, newIrNode];
        });
      } else {
        const newIrNode = new IrNode({
          name,
          coords: [tagStart, 0],
          children: [],
        });

        yield* Ref.update(irNodesState, (irNodes) => {
          return [...irNodes, newIrNode];
        });
      }

      tagStart = null;
      tagEnd = null;
      continue;
    }

    // No collection data, and no `textStart`, means we start
    // collecting text data
    if (tagStart === null && tagEnd === null && textStart === null) {
      textStart = i;
    }

    // If we have all text data, collect
    if (textStart !== null && textEnd !== null) {
      const newIrNode = new IrNode({
        name: "_text_node",
        coords: [textStart, textEnd],
        children: [],
      });

      yield* Ref.update(irNodesState, (irNodes) => {
        return [...irNodes, newIrNode];
      });

      textStart = null;
      textEnd = null;
    }
  }
});

/**
 * For given `irNodes`, finds all IR nodes within the given `coords`,
 * and returns back its index in state, as well as the object itself.
 */
const findIrNodesInCoords = (
  irNodes: IrNode[],
  coords: [number, number],
): [number, IrNode][] => {
  return irNodes.reduce(
    (nodes: [number, IrNode][], node: IrNode, i: number) => {
      if (node.coords[0] > coords[0] && node.coords[1] < coords[1]) {
        nodes.push([i, node]);
      }

      return nodes;
    },
    [],
  );
};

/**
 * An Effect which joins IR Nodes together - meaning that
 * it turns a flat list of IR Nodes into a tree of IR Nodes,
 * depending on if they have a parent or not, according to their coords.
 */
const joinIrNodesEffect = Effect.gen(function* () {
  const irNodesState = yield* IrNodesState;

  const joinIrNodes = (irNodes: IrNode[], added: Set<string>): IrNode[] => {
    return irNodes.reduce((iterNodes: IrNode[], node: IrNode) => {
      const nodeCoordsKey = JSON.stringify(node.coords);

      if (added.has(nodeCoordsKey)) return iterNodes;
      added.add(nodeCoordsKey);

      const iterNodeChildrenList = findIrNodesInCoords(irNodes, node.coords);
      const iterNodeChildrenNodes = iterNodeChildrenList.map((n) => n[1]);

      node.children = joinIrNodes(iterNodeChildrenNodes, added);
      iterNodes.push(node);

      return iterNodes;
    }, []);
  };

  yield* Ref.update(irNodesState, (irNodes) => {
    return joinIrNodes(irNodes, new Set<string>());
  });
});

/**
 * From given `html` and `coords` tries to return the attribute
 * part of a string. So for example if the coordinates point to a
 * string such as:
 *
 * ```html
 * <div class="some-class"></div>
 * ```
 *
 * Then it will attempt to find and return this:
 *
 * ```html
 * class="some-class"
 * ```
 */
const nodeAttrsStrFromCoords = (
  html: string,
  coords: [number, number],
): Option.Option<string> => {
  const nodeStr = html.substring(coords[0], coords[1]);
  let attrsStrStart: Option.Option<number> = Option.none();
  let attrsStrEnd: Option.Option<number> = Option.none();

  for (let i = 0; i < nodeStr.length; i++) {
    const char = nodeStr.at(i);

    if (char === ">") {
      attrsStrEnd = Option.some(i);
      break;
    }

    if (attrsStrStart === null && char === " ") {
      attrsStrStart = Option.some(i + 1);
    }
  }

  if (Option.isNone(attrsStrStart) || Option.isNone(attrsStrEnd)) {
    return Option.none();
  }

  return Option.some(nodeStr.substring(attrsStrStart.value, attrsStrEnd.value));
};

/**
 * For given `html` and `coords` tries to return the Attrs object. For example
 * if the coordinates point to a string such as:
 *
 * ```html
 * <div class="some-class"></div>
 * ```
 *
 * Then it will attempt to find and return this object:
 *
 * ```typescript
 * {
 *   class: "some-class"
 * }
 * ```
 */
const nodeAttrsFromCoords = (html: string, coords: [number, number]): Attrs => {
  let attributes: Attrs = {};
  const attrsStr = nodeAttrsStrFromCoords(html, coords);

  if (Option.isNone(attrsStr)) {
    return attributes;
  }

  let iterAttrName = "";
  let iterAttrValue: string | null = null;

  for (let i = 0; i < attrsStr.value.length; i++) {
    const char = attrsStr.value.at(i);

    // if we encounter a space, and the last char of `iterAttrValue` is `"`
    // it means we're not in an attr value, in which case a
    // space would be part of the value, but rather ending an attribute
    // declaration and moving onto the next one.
    if (
      char === " " &&
      iterAttrValue !== null &&
      iterAttrValue?.at(-1) !== '"'
    ) {
      if (iterAttrValue?.at(0) === '"' && iterAttrValue?.at(-1) === '"') {
        iterAttrValue = iterAttrValue?.substring(1, iterAttrValue?.length - 1);
      }

      attributes[`${iterAttrName}${char}`.trim()] = iterAttrValue;
      iterAttrName = "";
      iterAttrValue = null;
      continue;
    }

    // same as above is true when we are the last char of the entire `attrStr`,
    // in which case we are ending an attribute declaration.
    if (i === attrsStr.value.length - 1 && iterAttrValue !== null) {
      iterAttrValue += char;

      if (iterAttrValue.at(0) === '"' && iterAttrValue.at(-1) === '"') {
        iterAttrValue = iterAttrValue.substring(1, iterAttrValue.length - 1);
      }

      attributes[`${iterAttrName}`.trim()] = iterAttrValue;
      iterAttrName = "";
      iterAttrValue = null;
      continue;
    }

    // and, same as above is also true when we encounter a space and there is
    // no `iterAttrValue`, meaning it is a Truthy attribute, which needs
    // no explicit value.
    if (
      (char === " " || i === attrsStr.value.length - 1) &&
      iterAttrValue === null
    ) {
      attributes[`${iterAttrName}${char}`.trim()] = true;
      iterAttrName = "";
      iterAttrValue = null;
      continue;
    }

    // If we encounter the `=` char, it means we are done with `iterAttrName`,
    // and can move on to start creating the `iterAttrValue`.
    if (iterAttrValue === null && char === "=") {
      iterAttrValue = "";
      continue;
    }

    // in all other cases if we have already set `iterAttrValue`, keep on
    // collecting it.
    if (iterAttrValue !== null) {
      iterAttrValue += char;
      continue;
    }

    // or if we have not set `iterAttrValue`, keep on collecting `iterAttrName`.
    iterAttrName += char;
  }

  return attributes;
};

/**
 * Transforms an `irNode` to a `Node`.
 */
const irNodeToNode = (html: string, irNode: IrNode): Node => {
  if (irNode.name === "_text_node") {
    return new TextNode(html.substring(irNode.coords[0], irNode.coords[1]));
  }

  if (voidNames.includes(irNode.name.toLowerCase())) {
    return new VoidNode({
      name: irNode.name,
      attributes: nodeAttrsFromCoords(html, irNode.coords),
      children: [],
    });
  }

  return new Node({
    name: irNode.name,
    attributes: nodeAttrsFromCoords(html, irNode.coords),
    children: [],
  });
};

/**
 * An Effect which creates Node's from IrNode's. Essentially
 * turning crude oil into refined oil.
 */
const createNodesEffect = Effect.gen(function* () {
  const irNodesState = yield* IrNodesState;
  const nodesState = yield* NodesState;
  const html = yield* Ref.get(yield* HtmlState);

  const createNodes = (irNodes: IrNode[]): Node[] => {
    return irNodes.reduce((nodes: Node[], irNode: IrNode) => {
      if (!irNode.children.length) {
        nodes.push(irNodeToNode(html, irNode));
      } else {
        const node = irNodeToNode(html, irNode);
        node.children = createNodes(irNode.children);
        nodes.push(node);
      }

      return nodes;
    }, []);
  };

  yield* Ref.set(nodesState, createNodes(yield* Ref.get(irNodesState)));
});

/**
 * An Effect which puts all the various effects for composing nodes
 * together to then return the final result as a Node tree.
 */
const composeNodesEffect = Effect.gen(function* () {
  yield* createIrNodesEffect;
  yield* joinIrNodesEffect;
  yield* createNodesEffect;

  return yield* Ref.get(yield* NodesState);
});

/**
 * Transform the given `html` string into a tree of Node's.
 */
const nodes = (html: string) => {
  return Effect.runSync(
    composeNodesEffect.pipe(
      Effect.provideServiceEffect(HtmlState, Ref.make(html)),
      Effect.provideServiceEffect(IrNodesState, Ref.make<IrNode[]>([])),
      Effect.provideServiceEffect(NodesState, Ref.make<Node[]>([])),
    ),
  );
};

/**
 * Traverses the given tree of `nodes` for the purposes of manipulating
 * said tree of `nodes`. It takes a predicate function `pred` which gets
 * a single Node passed to it during traversal and which must return a
 * `Node` or `null`.
 *
 * If you return the same exact node, it will change nothing, but if you return
 * an updated node, it will update, and if you return a completely new node,
 * it will replace it. Returning `null` will remove the node from the tree.
 *
 * ---
 *
 * ### Types of nodes:
 *
 * - `Node` - generic node
 * - `VoidNode` - a self-closing node (has no children)
 * - `TextNode` - a node containing only plain text
 * - `FragmentNode` - a node which children replace itself
 */
const traverse = (nodes: Node[], pred: (node: Node) => Node | null): Node[] => {
  return nodes.reduce((updatedNodes: Node[], node: Node) => {
    const updatedNode = pred(node);

    // If the callback returns `null`, it means we want to remove the node.
    if (updatedNode === null) {
      return updatedNodes;
    }

    // If the callback returns a `TextNode`, we don't need to traverse its children.
    if (updatedNode instanceof TextNode) {
      updatedNodes.push(updatedNode);

      return updatedNodes;
    }

    // If the callback returns a `FragmentNode`, we need to traverse its children,
    // as it's children will replace the `FragmentNode` itself.
    if (updatedNode instanceof FragmentNode) {
      updatedNodes.push(...traverse(updatedNode.children, pred));

      return updatedNodes;
    }

    // If the callback returns a `VoidNode`, we don't need to traverse its children.
    if (updatedNode instanceof VoidNode) {
      updatedNodes.push(updatedNode);

      return updatedNodes;
    }

    // For all other nodes, we need to traverse its children.
    updatedNode.children = traverse(updatedNode.children, pred);
    updatedNodes.push(updatedNode);

    return updatedNodes;
  }, []);
};

/**
 * Very similar to the `traverse` function, except that instead of manipulating
 * the tree of nodes, it simply finds nodes matching the predicate function
 * passed as `pred` instead. Useful if you want to get only a subset of the
 * tree of nodes.
 */
const find = (nodes: Node[], pred: (node: Node) => boolean): Node[] => {
  return nodes.reduce((foundNodes: Node[], node: Node) => {
    if (pred(node)) {
      foundNodes.push(node);
    }

    if (node.children.length) {
      foundNodes.push(...find(node.children, pred));
    }

    return foundNodes;
  }, []);
};

/**
 * -----------------------------------------------------------------
 * Serializer
 * -----------------------------------------------------------------
 */
export type Serializer<T> = (nodes: Node[]) => T;

/**
 * Serializer the given `nodes` into an HTML string.
 */
const htmlSerializer: Serializer<string> = (nodes: Node[]): string => {
  return nodes.reduce((html: string, node: Node) => {
    if (node instanceof TextNode) {
      return `${html}${node.value}`;
    }

    if (node instanceof FragmentNode) {
      return `${html}${htmlSerializer(node.children)}`;
    }

    const attrsStr = Object.entries(node.attributes).reduce(
      (acc, [key, value]) => {
        if (value === true) {
          return `${acc} ${key}`;
        }

        return `${acc} ${key}="${value}"`;
      },
      "",
    );

    html += `<${node.name}${attrsStr}>`;

    if (!(node instanceof VoidNode)) {
      html += htmlSerializer(node.children);
      html += `</${node.name}>`;
    }

    return html;
  }, "");
};

/**
 * Serializes the given `nodes` into a JSON string.
 */
const jsonSerializer: Serializer<string> = (nodes: Node[]): string => {
  const reducer = (reduceNodes: Node[]) =>
    reduceNodes.reduce((data: Record<string, unknown>[], node: Node) => {
      if (node instanceof TextNode) {
        data.push({
          type: "textNode",
          value: node.value,
        });
      }

      if (node instanceof FragmentNode) {
        data.push({
          type: "fragmentNode",
          children: reducer(node.children),
        });
      }

      if (node instanceof VoidNode) {
        data.push({
          type: "voidNode",
          name: node.name,
          attributes: node.attributes,
          children: reducer(node.children),
        });
      }

      data.push({
        type: "node",
        name: node.name,
        attributes: node.attributes,
        children: reducer(node.children),
      });

      return data;
    }, []);

  return JSON.stringify(reducer(nodes));
};

/**
 * Serializes the given `nodes` with the given `serializer`.
 */
const serialize = <T>(nodes: Node[], serializer: Serializer<T>): T => {
  return serializer(nodes);
};

/**
 * -----------------------------------------------------------------
 * Export
 * -----------------------------------------------------------------
 */
export default {
  nodes,
  traverse,
  find,
  serialize,
  Node,
  TextNode,
  FragmentNode,
  VoidNode,
  Serializer: {
    Html: htmlSerializer,
    Json: jsonSerializer,
  },
};

File: src/dompa.rs
use serde::{Deserialize, Serialize, Serializer};
use std::collections::{BTreeMap, HashSet};

#[derive(Debug, Clone)]
struct IrNode {
    name: String,
    coords: (usize, usize),
    children: Vec<IrNode>,
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(untagged)]
pub enum NodeAttrVal {
    /// Represents a string HTML attribute, i.e:
    ///
    /// ```html
    /// <element attribute="value">
    /// ```
    String(String),

    /// Represents a truthy HTML attribute, i.e:
    ///
    /// ```html
    /// <element checked>
    /// ```
    True,
}

impl Serialize for NodeAttrVal {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        match self {
            NodeAttrVal::String(str) => serializer.serialize_str(str),
            NodeAttrVal::True => serializer.serialize_bool(true),
        }
    }
}

impl NodeAttrVal {
    /// Simplifies the creation of a `NodeAttrVal::String` variant by providing
    /// a shorthand function:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// NodeAttrVal::string("value");
    /// ```
    ///
    /// The verbose variant looks like this:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// NodeAttrVal::String(String::from("value"));
    /// ```
    pub fn string(value: impl Into<String>) -> Self {
        NodeAttrVal::String(value.into())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Node {
    Block(BlockNode),
    Text(TextNode),
    Void(VoidNode),
    Fragment(FragmentNode),
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BlockNode {
    pub name: String,
    pub attributes: BTreeMap<String, NodeAttrVal>,
    pub children: Vec<Node>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct TextNode {
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct VoidNode {
    pub name: String,
    pub attributes: BTreeMap<String, NodeAttrVal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FragmentNode {
    pub children: Vec<Node>,
}

// Adding some helper methods to the Node struct to make the usage a little
// less verbose.
impl Node {
    /// Simplifies the creation of a `BlockNode` by providing a shorthand
    /// function:
    ///
    /// ```rust
    /// use dompa::*;
    /// use std::collections::BTreeMap;
    ///
    /// Node::block("div", BTreeMap::new(), vec![]);
    /// ```
    ///
    /// The verbose variant looks like this:
    ///
    /// ```rust
    /// use dompa::*;
    /// use std::collections::BTreeMap;
    ///
    /// Node::Block(BlockNode {
    ///   name: String::from("div"),
    ///   attributes: BTreeMap::new(),
    ///   children: vec![]
    /// });
    /// ```
    pub fn block(
        name: impl Into<String>,
        attributes: BTreeMap<String, NodeAttrVal>,
        children: Vec<Node>,
    ) -> Self {
        Node::Block(BlockNode {
            name: name.into(),
            attributes,
            children,
        })
    }

    /// Simplifies the simplified shorthand function (`Node::block`) even more for when
    /// you don't want to set attributes, in which case you can simply call this:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// Node::simple_block("div", vec![]);
    /// ```
    pub fn simple_block(name: impl Into<String>, children: Vec<Node>) -> Self {
        Node::block(name, BTreeMap::new(), children)
    }

    /// Simplifies the creation of a `TextNode` by providing a shorthand
    /// function:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// Node::text("Hello, World!");
    /// ```
    ///
    /// The verbose variant looks like this:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// Node::Text(TextNode {
    ///   value: String::from("Hello, World!")
    /// });
    /// ```
    pub fn text(value: impl Into<String>) -> Self {
        Node::Text(TextNode {
            value: value.into(),
        })
    }

    /// Simplifies the creation of a `VoidNode` by providing a shorthand
    /// function:
    ///
    /// ```rust
    /// use dompa::*;
    /// use std::collections::BTreeMap;
    ///
    /// Node::void("img", BTreeMap::new());
    /// ```
    ///
    /// The verbose variant looks like this:
    ///
    /// ```rust
    /// use dompa::*;
    /// use std::collections::BTreeMap;
    ///
    /// Node::Void(VoidNode {
    ///   name: String::from("img"),
    ///   attributes: BTreeMap::new()
    /// });
    /// ```
    pub fn void(name: impl Into<String>, attributes: BTreeMap<String, NodeAttrVal>) -> Self {
        Node::Void(VoidNode {
            name: name.into(),
            attributes,
        })
    }

    /// Simplifies the simplifier shrothand function (`Node::void`) even more for when
    /// you don't want to set attributes, in which case you can simply call this:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// Node::simple_void("img");
    /// ```
    pub fn simple_void(name: impl Into<String>) -> Self {
        Node::void(name, BTreeMap::new())
    }

    /// Simplifies the creation of a `FragmentNode` by providing a shrothand
    /// function:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// Node::fragment(vec![]);
    /// ```
    ///
    /// The verbose variant looks like this:
    ///
    /// ```rust
    /// use dompa::*;
    ///
    /// Node::Fragment(FragmentNode {
    ///   children: vec![]
    /// });
    /// ```
    pub fn fragment(children: Vec<Node>) -> Self {
        Node::Fragment(FragmentNode { children })
    }
}

// Nodes which are automatically parsed as self-closing.
// TODO: should this be configurable?
static VOID_NODES: [&'static str; 14] = [
    "!doctype", "area", "base", "br", "col", "embed", "hr", "img", "input", "link", "meta",
    "source", "track", "wbr",
];

fn set_end_coord_of_last_ir_node(nodes: &mut [IrNode], name: &str, end_coord: usize) {
    if let Some(node) = nodes
        .iter_mut()
        .rev()
        .find(|node| node.name == name && node.coords.1 == 0)
    {
        node.coords = (node.coords.0, end_coord);
    }
}

fn create_ir_nodes(html: &str) -> Vec<IrNode> {
    let mut ir_nodes: Vec<IrNode> = Vec::from([]);
    let mut tag_start: Option<usize> = None;
    let mut tag_end: Option<usize> = None;
    let mut text_start: Option<usize> = None;
    let mut text_end: Option<usize> = None;

    for (i, char) in html.chars().enumerate() {
        // start of a tag or end of a text node
        if char == '<' {
            if text_start.is_some() {
                text_end = Some(i);
            }

            tag_start = Some(i);
        }
        // if we're at  the last char and it's not the end of a tag,
        // means w'ere in a text node
        else if i == html.len() - 1 && char != '>' {
            text_end = Some(i + 1);
        }
        // end of a tag
        else if char == '>' {
            tag_end = Some(i + 1);
        }

        // when we have both tag_start and tag_end, collect the tag
        if let (Some(start), Some(end)) = (tag_start, tag_end) {
            let tag = &html[start..end];
            let contents = &tag[1..tag.len() - 1];
            let contents_split: Vec<&str> = contents.split(" ").collect();
            let name = contents_split[0].to_string();
            let is_void_node = VOID_NODES.contains(&name.as_str());

            // if this is a closing tag, close the last node
            if tag.starts_with("</") {
                set_end_coord_of_last_ir_node(&mut ir_nodes, &name[1..], end);
            }
            // otherwise, business as usual, create the IrNode
            else {
                ir_nodes.push(IrNode {
                    name,
                    coords: (start, if is_void_node { end } else { 0 }),
                    children: Vec::new(),
                });
            }

            tag_start = None;
            tag_end = None;
            continue;
        }

        // no collection data, and no text_start, which means we start
        // collecting text data.
        if tag_start.is_none() && tag_end.is_none() && text_start.is_none() {
            text_start = Some(i);
        }

        // if we have all the text data, collect it
        if let (Some(start), Some(end)) = (text_start, text_end) {
            ir_nodes.push(IrNode {
                name: "_text_node".to_string(),
                coords: (start, end),
                children: Vec::new(),
            });

            text_start = None;
            text_end = None;
        }
    }

    ir_nodes
}

fn find_ir_nodes_in_coords(ir_nodes: &[IrNode], coords: (usize, usize)) -> Vec<IrNode> {
    ir_nodes
        .iter()
        .filter(|node| node.coords.0 > coords.0 && node.coords.1 < coords.1)
        .cloned()
        .collect()
}

fn join_ir_nodes_inner(nodes: &[IrNode], seen: &mut HashSet<(usize, usize)>) -> Vec<IrNode> {
    nodes
        .iter()
        .filter_map(|node| {
            // skip if we've seen it
            if !seen.insert(node.coords) {
                return None;
            }

            let children = find_ir_nodes_in_coords(nodes, node.coords);
            let processed_children = join_ir_nodes_inner(&children, seen);

            Some(IrNode {
                coords: node.coords,
                children: processed_children,
                ..node.clone()
            })
        })
        .collect()
}

fn join_ir_nodes(ir_nodes: Vec<IrNode>) -> Vec<IrNode> {
    join_ir_nodes_inner(&ir_nodes, &mut HashSet::new())
}

fn attrs_str_from_coords(html: &str, coords: (usize, usize)) -> Option<String> {
    let raw_tag = &html[coords.0..coords.1];
    let end = raw_tag.find('>')?;
    let start = raw_tag.find(|c: char| c.is_whitespace())? + 1;

    // start has to be before end, and we should actually have
    // attributes i.e start should not be equal to the tag length.
    if start >= end || start == raw_tag.len() {
        return None;
    }

    Some(raw_tag[start..end].to_string())
}

fn attrs_from_coords(html: &str, coords: (usize, usize)) -> BTreeMap<String, NodeAttrVal> {
    let mut attrs = BTreeMap::new();

    let Some(attrs_str) = attrs_str_from_coords(&html, coords) else {
        return attrs;
    };

    let mut chars = attrs_str.chars().peekable();
    let mut current_name = String::new();

    while let Some(c) = chars.next() {
        match c {
            // skip leading whitespace
            c if c.is_whitespace() && current_name.is_empty() => continue,

            // handle whitespace after attribute name or value
            c if c.is_whitespace() => {
                if !current_name.is_empty() {
                    attrs.insert(current_name.clone(), NodeAttrVal::True);
                    current_name.clear();
                }
            }

            // parse attribute value
            '=' => {
                if let Some('"') = chars.next() {
                    let value: String = chars.by_ref().take_while(|&c| c != '"').collect();

                    attrs.insert(current_name.clone(), NodeAttrVal::String(value));
                    current_name.clear();

                    // skip the space after the closing quote, if present
                    if let Some(' ') = chars.peek() {
                        chars.next();
                    }
                }
            }

            // collect attribute name
            _ => current_name.push(c),
        }
    }

    // handle last attribute, if it's a boolean one
    if !current_name.is_empty() {
        attrs.insert(current_name, NodeAttrVal::True);
    }

    attrs
}

fn create_nodes(html: &str, ir_nodes: Vec<IrNode>) -> Vec<Node> {
    ir_nodes
        .into_iter()
        .map(|ir_node| match ir_node.name.as_str() {
            "_text_node" => Node::Text(TextNode {
                value: html[ir_node.coords.0..ir_node.coords.1].to_string(),
            }),

            name if VOID_NODES.contains(&name) => Node::Void(VoidNode {
                name: ir_node.name,
                attributes: attrs_from_coords(html, ir_node.coords),
            }),

            _ => Node::Block(BlockNode {
                name: ir_node.name,
                attributes: attrs_from_coords(html, ir_node.coords),
                children: create_nodes(html, ir_node.children),
            }),
        })
        .collect()
}

/// Transforms a given `html` string into a node tree.
pub fn nodes(html: String) -> Vec<Node> {
    let ir_nodes = create_ir_nodes(&html);
    let joined_ir_nodes = join_ir_nodes(ir_nodes);

    return create_nodes(&html, joined_ir_nodes);
}

/// Traverses the given `nodes` node tree and returns an updated tree based
/// on `callable`. The `callable` must return a `Node` if you wish to either
/// replace the node or update it, and return `None` if you wish to remove it.
pub fn traverse<F>(nodes: Vec<Node>, mut callable: F) -> Vec<Node>
where
    F: for<'r> FnMut(&'r Node) -> Option<Node>,
{
    traverse_inner(nodes, &mut callable)
}

fn traverse_inner<F>(nodes: Vec<Node>, callable: &mut F) -> Vec<Node>
where
    F: for<'r> FnMut(&'r Node) -> Option<Node>,
{
    let mut result = Vec::new();

    for node in nodes {
        if let Some(t) = callable(&node) {
            match t {
                Node::Block(mut blk) => {
                    blk.children = traverse_inner(blk.children, callable);
                    result.push(Node::Block(blk));
                }
                Node::Fragment(frag) => {
                    result.extend(traverse_inner(frag.children, callable));
                }
                Node::Text(_) | Node::Void(_) => result.push(t),
            }
        }
    }

    result
}

fn attrs_to_html_str(attrs: BTreeMap<String, NodeAttrVal>) -> String {
    attrs.into_iter().fold(String::new(), |acc, attr| {
        let html = match attr {
            (key, NodeAttrVal::True) => key,
            (key, NodeAttrVal::String(val)) => format!("{}=\"{}\"", key, val),
        };

        format!("{} {}", acc, html).to_string()
    })
}

/// Transform the given `nodes` node tree into a HTML string.
pub fn to_html(nodes: Vec<Node>) -> String {
    nodes.into_iter().fold(String::new(), |acc, node| {
        let html = match node {
            // Block nodes
            Node::Block(block_node) => {
                format!(
                    "<{name}{attrs}>{content}</{name}>",
                    name = block_node.name,
                    attrs = attrs_to_html_str(block_node.attributes),
                    content = to_html(block_node.children),
                )
            }

            // Text nodes
            Node::Text(text_node) => text_node.value,

            // Void nodes
            Node::Void(void_node) => {
                format!(
                    "<{name}{attrs}>",
                    name = void_node.name,
                    attrs = attrs_to_html_str(void_node.attributes)
                )
            }

            // Fragment nodes
            Node::Fragment(fragment_node) => to_html(fragment_node.children),
        };

        format!("{}{}", acc, html)
    })
}

/// Transform the given `nodes` node tree into a JSON string.
pub fn to_json(nodes: Vec<Node>) -> String {
    serde_json::to_string(&nodes).unwrap_or(String::from("{}"))
}
